### DISCUSSION TITLE
cattrs >=22.1.0 requires exceptiongroup package, which has not been released

### DISCUSSION THREAD

**Discussion by xylar** [apache/airflow#25933] [General] [answered]
This isn't exactly a bug but it's a pretty big impediment to making a 2.3.4 package on conda-forge.  See
https://github.com/conda-forge/airflow-feedstock/pull/91
The latest release requires `cattrs >=22.1.0`, which depends on `exceptiongroup`:
https://github.com/python-attrs/cattrs/blob/v22.1.0/pyproject.toml#L27
The problem is that `exceptiongroup` hasn't yet been released:
https://github.com/agronholm/exceptiongroup/issues/20
and likely won't be until python 3.11 is released in October
https://www.python.org/downloads/release/python-3110rc1/#:~:text=This%20release%2C%203.11.&text=The%20second%20candidate%20and%20the,%2C%202022%2D10%2D03

It would be possible to build Airflow 2.3.4 if users included the flag `-c conda-forge/label/exceptiongroup_dev` when they try to install `apache-airflow` from conda-forge but this isn't a very intuitive requirement.

For now, I think we will not be able to build a 2.3.4 package on conda-forge.  I wanted to bring this up here in case there are suggestions for how to work around this problem.

**Comment by potiuk** [comment, marked as answer]
I suggest to fix conda @xylar . Conda in this respect is not PEP 440 compliant (and `pip` is).

The reason why exceptiongroup is installed and cattrs can be installed when airflow is installed with `pip` is because pip correctly implement `PEP 440` https://peps.python.org/pep-0440/#handling-of-pre-releases

> Pre-releases of any kind, including developmental releases, are implicitly excluded from all version specifiers, unless they are already present on the system, explicitly requested by the user, or **if the only available version that satisfies the version specifier is a pre-release**.

> By default, dependency resolution tools SHOULD:

> * accept already installed pre-releases for all version specifiers
> * accept remotely available pre-releases for version specifiers **where there is no final or post release that satisfies the version specifier**
> *  exclude all other pre-releases from consideration

The second point is why `pip` install airflow with ``exceptiongroup = 1.0.0rc8``

cc: @uranusjr - I believe I understood it correctly.

**Comment by xylar** [comment]
@potiuk, thanks for your answer.  I also got a similar answer from core folks from conda-forge (pointing out https://github.com/conda-forge/cfep/blob/main/cfep-05.md#exceptions).  This is an example of the kind of knowledge that sometimes takes awhile to percolate through an all-volunteer community of maintainers like on conda-forge and I just hadn't come across this before.  Now that I know, it seems easy enough to fix.

Thanks agian!

**Reply by potiuk** [reply]
Cool :)

**Comment by uranusjr** [comment]
Regardless of package manager support, releasing a “stable” software with pre-release-only dependencies feels like a somewhat suboptimal approach to run a project to me, but I guess people have different opinions.

**Reply by xylar** [reply]
My sense talking with @agronholm is that it doesn't really make sense to release `exceptiongroup` until python 3.11 is out because then they just have to do another release afterward -- it's a moving target.
https://github.com/agronholm/exceptiongroup/issues/20
At first, I tended to agree with you that this didn't seem like a good policy but it does seem like it's not that unusual now that I've looked into it a bit further.

**Reply by potiuk** [reply]
Yeah. It is suboptimal - I think we all agree :) . But it's also "fact-of-life" it happens and even been acknowledged by PEP 440 authors :D
